tests/test_coffee_optimise.py

```diff
@@ -0,0 +1,112 @@
+from __future__ import absolute_import, print_function, division
+
+import pytest
+
+from gem.gem import Index, Indexed, Product, Variable, Division, Literal, Sum
+from gem.optimise import replace_division, reassociate_product
+from tsfc.coffee_mode import optimise_expressions, count_flop
+
+
+def test_replace_div():
+    i = Index()
+    A = Variable('A', ())
+    B = Variable('B', (6,))
+    Bi = Indexed(B, (i,))
+    d = Division(Bi, A)
+    result = replace_division([d])[0]
+
+    assert isinstance(result, Product)
+    assert isinstance(result.children[1], Division)
+
+
+def test_reassociate_product():
+    """Test recursive reassociation of products according to the rank of the
+    factors. For example: ::
+
+        C[i,j]*(D[i,j,k] + Q[i,j]*t*P[i])*A*B[i]
+
+    becomes: ::
+
+        (D[i,j,k] + Q[i,j]*(t*P[i]))*(C[i,j]*(A*B[i]))
+
+    """
+    i = Index()
+    j = Index()
+    k = Index()
+    A = Variable('A', ())
+    Bi = Indexed(Variable('B', (6,)), (i,))
+    Cij = Indexed(Variable('C', (6, 6)), (i, j))
+    Dijk = Indexed(Variable('D', (6, 6, 6)), (i, j, k))
+    Pi = Indexed(Variable('P', (6,)), (i,))
+    Qij = Indexed(Variable('Q', (6, 6)), (i, j))
+    t = Literal(8)
+    p = Product(Product(Product(Cij, Sum(Dijk, Product(Product(Qij, t), Pi))), A), Bi)
+    result = reassociate_product([p])[0]
+    # D[i,j,k] + t*P[i]*Q[i,j]
+    assert isinstance(result.children[0], Sum)
+    # C[i,j]
+    assert result.children[1].children[0] == Cij
+    # A * B[i]
+    assert result.children[1].children[1] == Product(A, Bi)
+    # t * P[i]
+    assert result.children[0].children[1].children[1] == Product(t, Pi)
```

Comment:
Would it be better to just assert that the result is equal to the expected expression?